src/databricks/sql/auth/oauth.py

```diff
@@ -0,0 +1,286 @@
+    oauth_response = json.loads(response.text)
+    return oauth_response
+
+
+def send_refresh_token_request(hostname, client_id, refresh_token):
+    idp_url = get_idp_url(hostname)
+    oauth_config = fetch_well_known_config(idp_url)
+    token_request_url = oauth_config['token_endpoint']
+    client = get_client(client_id=client_id)
+    token_request_body = client.prepare_refresh_body(
+        refresh_token=refresh_token, client_id=client.client_id)
+    return send_token_request(token_request_url, token_request_body)
+
+
+def get_tokens_from_response(oauth_response):
+    access_token = oauth_response['access_token']
+    refresh_token = oauth_response['refresh_token'] if 'refresh_token' in oauth_response else None
+    return access_token, refresh_token
+
+
+def check_and_refresh_access_token(hostname, access_token, refresh_token):
+    now = datetime.now(tz=UTC)
+    # If we can't decode an expiration time, this will be expired by default.
+    expiration_time = now
+    try:
+        # This token has already been verified and we are just parsing it.
+        # If it has been tampered with, it will be rejected on the server side.
+        # This avoids having to fetch the public key from the issuer and perform
+        # an unnecessary signature verification.
+        decoded = jwt.decode(access_token, options={"verify_signature": False})
+        expiration_time = datetime.fromtimestamp(decoded['exp'], tz=UTC)
+    except PyJWTError as err:
+        error_and_quit(err)
+
+    if expiration_time > now:
+        # The access token is fine. Just return it.
+        return access_token, refresh_token, False
+
+    if not refresh_token:
+        error_and_quit("OAuth access token expired on {expiration_time}."
+                       .format(expiration_time=expiration_time))
+
+    # Try to refresh using the refresh token
+    click.echo("Attempting to refresh OAuth access token that expired on {expiration_time}"
+               .format(expiration_time=expiration_time))
+    oauth_response = send_refresh_token_request(hostname, refresh_token)
+    fresh_access_token, fresh_refresh_token = get_tokens_from_response(oauth_response)
+    return fresh_access_token, fresh_refresh_token, True
+
+
+def get_tokens(hostname, client_id, scope=None):
+    idp_url = get_idp_url(hostname)
+    oauth_config = fetch_well_known_config(idp_url)
+    # We are going to override oauth_config["authorization_endpoint"] use the
+    # /oidc redirector on the hostname, which may inject additional parameters.
+    auth_url = "{}oidc/v1/authorize".format(hostname)
+    state = token_urlsafe(16)
+    (verifier, challenge) = get_challenge()
+    client = get_client(client_id)
+    redirect_url = get_redirect_url()
+    try:
+        auth_response = get_authorization_code(
+            client,
+            auth_url,
+            redirect_url,
+            scope,
+            state,
+            challenge,
+            REDIRECT_PORT)
+    except OAuth2Error as err:
```

Comment:
```suggestion
    except OAuth2Error as e:
```

is more idiomatic way of catching exceptions in python

Reply:
thanks addressed.